### PR TITLE
Fixed default MSVS build

### PR DESCRIPTION
**Pull request by notecola**
Fixed Windows compilation errors. Should work now for MS VC 2015 and newer.

### CODE REVIEW

File: ggml.h [hunk 1]
```diff
@@ -2217,8 +2217,8 @@ extern "C" {
     // Internal types and functions exposed for tests and benchmarks
     //
 
-#ifdef  __cplusplus
-// restrict not standard in C++
+#if defined(__cplusplus) || defined(_MSC_VER)
+// GGML_RESTRICT not standard in C++
```
Comment (cebtenzzre) [inline ggml.h:2221]: Looks like a bad find-and-replace - the comment should stay as it was.

File: examples/main/main.cpp.bak
Comment (cebtenzzre) [inline examples/main/main.cpp.bak:1]: Delete this file.

File: examples/infill/infill.cpp.bak
Comment (cebtenzzre) [inline examples/infill/infill.cpp.bak:1]: Delete this file.

File: examples/main/main.cpp [hunk 2]
```diff
@@ -392,10 +399,7 @@ int main(int argc, char ** argv) {
         sigint_action.sa_flags = 0;
         sigaction(SIGINT, &sigint_action, NULL);
 #elif defined (_WIN32)
-        auto console_ctrl_handler = +[](DWORD ctrl_type) -> BOOL {
-            return (ctrl_type == CTRL_C_EVENT) ? (sigint_handler(SIGINT), true) : false;
-        };
-        SetConsoleCtrlHandler(reinterpret_cast<PHANDLER_ROUTINE>(console_ctrl_handler), true);
+		SetConsoleCtrlHandler(reinterpret_cast<PHANDLER_ROUTINE>(ConsoleCtrlHandler), TRUE);
```
Comment (cebtenzzre) [inline examples/main/main.cpp:402]: Use spaces, not tabs. Also, what was wrong with the lambda?

File: CMakeLists.txt [hunk 3]
```diff
@@ -500,6 +500,9 @@ if (LLAMA_CUBLAS)
 endif()
 
 if (WIN32)
+    if (MSVC)
+        set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /bigobj")
```
Comment (cebtenzzre) [inline CMakeLists.txt:504]: There should be a comment explaining why this is needed. Presumably object file size limitations?

**Comment by ggerganov** [comment]
How come we never had problems with MSVC until now? I don't use Windows, but I thought the project builds nicely

File: examples/main/main.cpp (hunk 2 above)
Comment (notecola) [reply examples/main/main.cpp]: MSVC 2015 cannot handle + in lambda. More info: https://stackoverflow.com/questions/26220448/operator-on-lambda

**Comment by bullno1** [comment]
FWIW, I don't have a problem on MSVC 2022. 

But I only do "warning as error" on my own code not dependencies.

**Comment by cebtenzzre** [comment]
> How come we never had problems with MSVC until now?

As for the restrict specifier and the plus operator on a lambda, it seems to be because VS 2015 is not fully C99 or C++11 compliant.

@notecola What is your use case for this old version of MSVC?

**Comment by notecola** [comment]
@cebtenzzre I have no use case besides studying the code. Not sure how many people are still using vc 2015. You may want to fix the usage of _restrict_ anyway since there is already _GGML_RESTRICT_ defined.

**Comment by cebtenzzre** [comment]
> You may want to fix the usage of _restrict_ anyway since there is already _GGML_RESTRICT_ defined.

But GGML_RESTRICT is only supposed to be needed for C++, where `restrict` is not standard. This project expects C11 which guarantees `restrict`.

**Comment by notecola** [comment]
Yes I meant some files have _GGML_RESTRICT_ and some _restrict_ so the point is about consistency.

**Comment by cebtenzzre** [comment]
> Yes I meant some files have _GGML_RESTRICT_ and some _restrict_ so the point is about consistency.

It's consistent on master - ggml.h needs GGML_RESTRICT because it uses `restrict` when included by a C source file but not with C++. Standard-conforming compilers have no need for such a macro outside of headers.

**Comment by notecola** [comment]
I see, thanks. What about ggml-quants.h?

**Comment by slaren** [comment]
`ggml-quants.h` is an internal header to ggml and should only be included in C code.

**Comment by notecola** [comment]
@slaren Got it, thank you.

**Pull request closed without merging**